File: src/main/scala/s/decorator/inheritance/Display.scala
package s.decorator.inheritance

object Display {

  def ofString(string: String): Display = new StringDisplay(string)

  def ofSideBorder(display: Display, ch: Char): Display =
    new SideBorder(display, ch)

  def ofFullBorder(display: Display): Display = new FullBorder(display)

}

abstract class Display {

  def columns: Int

  def rows: Int

  def rowText(row: Int): String

  def show(): Unit = {
    for (i <- 0 until rows) {
      println(rowText(i))
    }
  }

}

File: src/main/java/j/decorator/inheritance/Display.java
package j.decorator.inheritance;

public abstract class Display {
  public static Display ofString(String string) {
    return new StringDisplay(string);
  }

  public static Display ofSideBorder(Display display, char ch) {
    return new SideBorder(display, ch);
  }

  public static Display ofFullBorder(Display display) {
    return new FullBorder(display);
  }

  public abstract int getColumns();

  public abstract int getRows();

  public abstract String getRowText(int row);

  public void show() {
    for (int i = 0; i < getRows(); i++) {
      System.out.println(getRowText(i));
    }
  }
}
